webapp/components/notifications/NotificationMenu/NotificationMenu.vue

```diff
@@ -46,24 +48,53 @@ export default {
           variables,
         })
         if (!(markAsRead && markAsRead.read === true)) return
-        this.notifications = this.notifications.map(n => {
-          return n.from.id === markAsRead.from.id ? markAsRead : n
+        this.displayedNotifications = this.displayedNotifications.map(n => {
+          return this.equalNotification(n, markAsRead) ? markAsRead : n
         })
       } catch (err) {
-        throw new Error(err)
+        this.$toast.error(err.message)
       }
     },
+    equalNotification(a, b) {
+      return a.from.id === b.from.id && a.createdAt === b.createdAt && a.reason === b.reason
+    },
   },
   computed: {
-    totalNotifications() {
-      return (this.notifications || []).length
+    notificationsCount() {
+      return (this.displayedNotifications || []).length
+    },
+    unreadNotificationsCount() {
+      let countUnread = 0
+      if (this.displayedNotifications) {
+        this.displayedNotifications.forEach(notification => {
+          if (!notification.read) countUnread++
+        })
+      }
+      return countUnread
     },
   },
   apollo: {
     notifications: {
       query() {
         return notificationQuery(this.$i18n)
       },
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-26T07:21:53Z" title="Thursday, September 26th 2019, 9:21:53 am +02:00">Sep 26, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -29,13 +41,37 @@ export default {
   },
   data() {
     return {
+      displayedNotifications: [],
       notifications: [],
     }
   },
   props: {
     placement: { type: String },
   },
+  watch: {
+    notifications: {
+      immediate: true,
+      handler(lastQueriedNotifications) {
+        if (lastQueriedNotifications && lastQueriedNotifications.length > 0) {
+          lastQueriedNotifications.forEach(updatedListNotification => {
+            const sameNotification = this.displayedNotifications.find(oldListNotification => {
+              return this.equalNotification(oldListNotification, updatedListNotification)
+            })
+            if (!sameNotification) this.displayedNotifications.unshift(updatedListNotification)
+          })
+        }
+      },
+    },
+  },
+
   methods: {
+    async updateNotifications() {
+      try {
+        this.$apollo.queries.notifications.refetch()
+      } catch (err) {
+        throw new Error(err)
+      }
+    },
```


I would prefer to have this callback deleted. I believe it creates bad user experience, see above :point_up:

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-09-26T08:55:55Z" title="Thursday, September 26th 2019, 10:55:55 am +02:00">Sep 26, 2019</time>_

---


Why bad user experience?
Because of the possible error?
So I shall use `this.$apollo.queries.notifications.refetch()` directly?
Or get rid of the update at all? Like you said …

I don’t agree with the last, like I said above …

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-26T13:01:54Z" title="Thursday, September 26th 2019, 3:01:54 pm +02:00">Sep 26, 2019</time>_

---


Bad user experience because of the situation where you are about to click on a notification but in the very moment when you want to click on it, the list gets updated and you suddenly click on sth. else.

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-09-27T09:09:31Z" title="Friday, September 27th 2019, 11:09:31 am +02:00">Sep 27, 2019</time>_

---


I implemented now two new things. See my general comment of the changes 👇🏼